pkg/hubble/filters/patterns.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2020 Authors of Hubble
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filters
+
+import (
+	"errors"
+	"regexp"
+	"strings"
+)
+
+var (
+	// fqdnRegexpStr matches an FQDN.
+	// FIXME this should not match components that begin or end with hyphens, e.g. -foo-
+	fqdnRegexpStr = `(?:[\-0-9a-z]+(?:\.[\-0-9a-z]+)*)`
+	_             = regexp.MustCompile(fqdnRegexpStr) // compile regexp to ensure that it is valid
+
+	errEmptyPattern                  = errors.New("empty pattern")
+	errInvalidRuneInPattern          = errors.New("invalid rune in pattern")
+	errMultipleTrailingDotsInPattern = errors.New("multiple trailing dots in pattern")
+	errTooManySlashesInPattern       = errors.New("too many slashes in pattern")
+)
+
+// canonicalizeFQDNPattern canonicalizes fqdnPattern by trimming space, trimming
+// up to one trailing dot, and converting it to lowercase.
+func canonicalizeFQDNPattern(fqdnPattern string) string {
+	fqdnPattern = strings.TrimSpace(fqdnPattern)
+	fqdnPattern = strings.TrimSuffix(fqdnPattern, ".")
+	fqdnPattern = strings.ToLower(fqdnPattern)
+	return fqdnPattern
+}
+
+// appendFQDNPatternRegexp appends the regular expression equivalent to
+// fqdnPattern to sb.
+func appendFQDNPatternRegexp(sb *strings.Builder, fqdnPattern string) error {
+	fqdnPattern = canonicalizeFQDNPattern(fqdnPattern)
+	switch {
+	case fqdnPattern == "":
+		return errEmptyPattern
+	case strings.HasSuffix(fqdnPattern, "."):
+		return errMultipleTrailingDotsInPattern
+	}
+	for _, r := range fqdnPattern {
+		switch {
+		case r == '.':
+			sb.WriteString(`\.`)
+		case r == '*':
+			sb.WriteString(`[\-\.0-9a-z]*`)
+		case r == '-':
+			fallthrough
+		case '0' <= r && r <= '9':
+			fallthrough
+		case 'a' <= r && r <= 'z':
+			sb.WriteRune(r)
+		default:
+			return errInvalidRuneInPattern
```

Comment:
`fmt.Errorf(...)` with the actual invalid rune would produce a better error message I think

Reply:
Done.